### DISCUSSION TITLE
Running model on cloud

### DISCUSSION THREAD

**Discussion by uturuncoglu** [ufs-community/ufs-weather-model#2011] [Q&A]
As a part of the UFS Coastal Project we are aiming to run the ufs-coastal (it is a fork of ufs-wearther-model) on cloud (specifically ParallelWorks) and creating some application level workflow to run some tests over there. I just wonder about the current status to use spack-stack in the could environment and running model over there. If there is a documentation that has detailed instructions that would be great.

**Comment by uturuncoglu** [comment]
@DomHeinzeller @BrianCurtis-NOAA Any suggestion?

**Reply by climbfuji** [reply]
See https://spack-stack.readthedocs.io/en/latest/PreConfiguredSites.html#noaa-parallel-works-aws-azure-gcloud - we have ready-to-go site configs that we (JCSDA and EPIC) use to build the unified environment on Parallel Works for every spack-stack release.

**Reply by uturuncoglu** [reply]
@climbfuji I don't have any experience of using parallel works but Is this available when you have instance? So, it seems we don't need to install the dependencies from scratch right? But I think we could install additional dependencies once we activate the environment.

**Reply by uturuncoglu** [reply]
@climbfuji BTW, thanks for your help. That is really great to have those information.

**Reply by climbfuji** [reply]
Unless you are a member of one of our projects, you'll have to build spack-stack yourself. But given the fact that the site config is maintained and known to work, it's a matter of copy and paste and then a long wait, because the `/contrib` filesystem is slow.

**Reply by uturuncoglu** [reply]
@climbfuji Is it possible to deploy your own docker container to work under Parallel Works? If so, I think preconfigured core container with the all dependencies will simplify the process of running the model and we could get rid of installing same things. Maybe you already do it under your project accounts. Not sure. BTW, Sorry about my lack of understanding about could and it is way of running.

**Reply by climbfuji** [reply]
I guess it's possible, but I haven't tried and it's not on our (JCSDA) priority list, so you'll have to figure it out yourself, sorry.

**Reply by uturuncoglu** [reply]
@climbfuji That is totally fine. Even if knowing it is possible is nice for me and we could explore that idea of using pre configured container under cloud. The ultimate aim is to use the cloud for ufs-coastal application level testing. Thanks again for your help.

**Reply by pvelissariou1** [reply]
@climbfuji Thank you fgor this information. I am wondering, is it possible to create a zipped tar archive of your spack-stack installation and share it with us so we can deploy it in our clusters (/contrib folder) in Parallel Works? This will be of great service if it can be done.

**Reply by uturuncoglu** [reply]
@pvelissariou1 That could work for now and testing but once spack-stack is updated this zip file needs to be updated too. I think if we could manage to have our own container created automatically (integrated with GiHub action in the application/workflow side) when new version is available. Then, we don't need to worry about manual work like copping etc. Also the container could be used by the community to test some thing in their environments.

**Reply by climbfuji** [reply]
What I've done in the past is to create a spack binary cache on an S3 bucket that can be used to install much faster. Whatever support you need is probably a task for EPIC, therefore I suggest you ask @natalie-perlin @ulmononian @RatkoVasic-NOAA for help.

**Reply by pvelissariou1** [reply]
@climbfuji @uturuncoglu Thank you both. @uturuncoglu The container is an excellent solution. On the same time, I would like to have spack-stack on our Parallel Works clusters as well to be used by other projects. In the container, the only compiler that we can have is gcc/gfortran compilers.

**Reply by pvelissariou1** [reply]
@natalie-perlin @ulmononian @RatkoVasic-NOAA We would like to have spack-stack installed (or an installation shared with us) in our ParallelWorks clusters (will be installed /contrib) so we can use it for all our model simulations. I am wondering what will be the easiest way to achieve this. Your help on this is highly appreciated.

**Reply by ulmononian** [reply]
@pvelissariou1 @uturuncoglu are you using a NOAA Parallelworks instance by chance (if so, gcp aws or azure)? we maintain the spack-stack on the EPIC contrib space; e.g., spack-stack/1.5.0 can be found here `/contrib/EPIC/spack-stack/spack-stack-1.5.0/envs/unified-env/install/modulefiles/Core`. please let me know if you are able to access this location from your space (we've had some trouble sharing contrib spaces between projects in the past). if you cannot, we can look at some other options (e.g. your team could utilize one of the spack-stack cloud configs such as https://github.com/JCSDA/spack-stack/tree/develop/configs/sites/noaa-aws, if using aws, for example).

**Reply by pvelissariou1** [reply]
@ulmononian Thank you for the reply. The issue is that we cannot access spack-stack from your space. At this point, I thought to deploy spack-stack in our PW clusters if you could share a tar archive of your spack-stack installation. I believe that somehow spack-stack should be available on PW for all cluster spaces. By the way, I checked spack-stack on hera, orion and hercules using UFS-Coastal and other modeling applications and everything works as expected.

**Reply by climbfuji** [reply]
The issue with sharing a tar file is that spack-stack by default optimizes for the CPU you are using. For instance, JCSDA uses hpc6id / c6i instances with a fairly recent Intel processor. EPIC wasn't able to use the spack-stack binaries we generated, because they are using a different/older instance family with a different CPU type that doesn't understand AVX-512 for example. Your safest bet would be to install yourselves using the site configs we maintain. Or to coordinate with EPIC to use the same instance type they do and then obtain their tar files.

**Reply by pvelissariou1** [reply]
@climbfuji Thank you, I will try to coordinate with EPIC on this.

**Reply by ulmononian** [reply]
@pvelissariou1 since you don't have access to either of jcsda/epic space, and given what @climbfuji described about tarballs/machine architecture, i think it could be best to install spack-stack from scratch in your space using one of our site configs for the noaa cloud (e.g., aws: https://github.com/JCSDA/spack-stack/tree/develop/configs/sites/noaa-aws) and followhttps://spack-stack.readthedocs.io/en/latest/CreatingEnvironments.html. you may have to edit the `compilers.yaml` and `packages.yaml` based on the compilers and external packages on your instance.

```

**Reply by pvelissariou1** [reply]
It seems compiling from scratch is a great option. First, I will try to coordinate on this with EPIC/UFS because we are developing UFS-Coastal based on UFS weather model and we want to be close with them in terms of development and testing. If this doesn't work, then I will build spack-stack from scratch per your recommendations.

**Comment by gspetro-NOAA** [comment]
@uturuncoglu Just wanted to check whether this was resolved offline with EPIC folks (in which case I'll close the discussion) or whether your team needs additional assistance.

**Comment by uturuncoglu** [comment]
@gspetro-NOAA I have no information about it but maybe @pvelissariou1 has. I personally did not part of discussion with the EPIC team. @pvelissariou1 Do you have any update?

**Comment by gspetro-NOAA** [comment]
@pvelissariou1 I will close this discussion since you seem to have determined a path forward, but feel free to open another discussion or email support.epic@noaa.gov if you have any further questions.
